File: euler/2.Even-Fibonacci-numbers/solution.lua
#!/usr/bin/env lua

function receive(prod)
   local good, value = coroutine.resume(prod)
   return value
end

function send(x)
   coroutine.yield(x)
end

function producer()
   return coroutine.create(
      function ()
         local a, b = 1, 2;
         send(a)
         send(b)
         while true do
            a, b = b, a + b
            send(b)
         end
   end)
end

function filter(prod)
   return coroutine.create(
      function()
         while true do
            local x = receive(prod)
            if x % 2 == 0 then
               send(x)
            end
         end
   end)
end

function consumer(prod, max)
   local sum = 0
   local x = receive(prod)
   while x < max do
      sum = sum + x
      x = receive(prod)
   end
   return sum
end

-- Fibonacci numbers: 1, 2, 3, 5, 8, 13, 21, 34, 55, 89, ...
assert(consumer(filter(producer()), 100) == 2 + 8 + 34)
print(consumer(filter(producer()), 4000000))

File: euler/2.Even-Fibonacci-numbers/solution.c
#include <assert.h>
#include <stdio.h>

int sum_even_fibonacci(int max)
{
     int pair[2] = {1, 2};
     int sum = 0;
     int last = 1;
     while (pair[last] <= max) {
          if (pair[last] % 2 == 0) {
               sum += pair[last];
          }
          last = !last;
          pair[last] = pair[0] + pair[1];
     }
     return sum;
}


int main(int argc, char* argv[])
{
     /* Fibonacci numbers: 1, 2, 3, 5, 8, 13, 21, 34, 55, 89, ... */
     assert(sum_even_fibonacci(100) == 2 + 8 + 34);
     printf("%d\n", sum_even_fibonacci(4000000));
}
